lib/jetstream/api/kv.ex

```diff
@@ -184,6 +184,69 @@ defmodule Jetstream.API.KV do
     end
   end
 
+  @doc """
+  Watches a bucket for addition and removal. You supply a handler that will receive add and
+  remove messages.
+
+  ## Examples
+
+      iex>{:ok, _watchres} = KV.watch(:gnat, "my_bucket", fn action, key, value -> IO.puts(action) end)
+  """
+  @type watch_handler :: (:key_added | :key_removed, String.t(), binary() -> nil)
+  @spec watch(conn :: Gnat.t(), bucket_name :: binary(), handler :: watch_handler()) ::
+          {:ok, {pid(), String.t()}} | {:error, any}
+  def watch(conn, bucket_name, handler) do
+    stream = stream_name(bucket_name)
+    inbox = Util.reply_inbox()
+    consumer_name = "all_key_values_watcher_#{Util.nuid()}"
+
+    {:ok, int_handler} =
+      Task.start(fn ->
+        receive_watched_keys(bucket_name, handler)
+      end)
+
+    with {:ok, sub} <- Gnat.sub(conn, int_handler, inbox),
+         {:ok, _consumer} <-
+           Consumer.create(conn, %Consumer{
+             durable_name: consumer_name,
+             deliver_subject: inbox,
+             stream_name: stream,
+             ack_policy: :none,
+             max_ack_pending: -1,
+             max_deliver: 1
+           }) do
+      {:ok, {sub, consumer_name}}
+    end
+  end
+
+  @doc """
+  Stops watching a given bucket. Supply the results of the watch function to stop it
+  """
+  def unwatch(conn, bucket_name, {sub_pid, consumer_name}) do
+    stream = stream_name(bucket_name)
```

Comment:
we also need to stop the `Task` here

Reply:
I agree, it is a bit onerous to pass all 3 of those around. going to think about this.

Reply:
So, since this is a long-lived process with state we want to keep track of I think we need to make it a GenServer. Something like this:

```elixir
defmodule Jetstream.API.KV.Watcher do
  use GenServer

  alias Jetstream.API.{Consumer, KV, Util}

  def start_link(opts) do
    GenServer.start_link(__MODULE__, opts, name: Keyword.get(opts, :name, __MODULE__))
  end

  def stop(pid) do
    GenServer.stop(pid)
  end

  def init(opts) do
    {:ok, {sub, consumer_name}} = subscribe(opts[:conn], opts[:bucket_name])

    {:ok,
     %{
       handler: opts[:handler],
       conn: opts[:conn],
       bucket_name: opts[:bucket_name],
       sub: sub,
       consumer_name: consumer_name
     }}
  end

  def terminate(_reason, state) do
    stream = KV.stream_name(state.bucket_name)

    :ok = Gnat.unsub(state.conn, state.sub)
    :ok = Consumer.delete(state.conn, stream, state.consumer_name)
  end

  def handle_info({:msg, %{topic: key, body: body, headers: headers}}, state) do
    key = KV.subject_to_key(key, state.bucket_name)

    if {"kv-operation", "DEL"} in headers do
      state.handler.(:key_deleted, key, body)
    end

    {:noreply, state}
  end

  def handle_info({:msg, %{topic: key, body: body}}, state) do
    key = KV.subject_to_key(key, state.bucket_name)

    state.handler.(:key_added, key, body)
    {:noreply, state}
  end

  defp subscribe(conn, bucket_name) do
    stream = KV.stream_name(bucket_name)
    inbox = Util.reply_inbox()
    consumer_name = "all_key_values_watcher_#{Util.nuid()}"

    with {:ok, sub} <- Gnat.sub(conn, self(), inbox),
         {:ok, _consumer} <-
           Consumer.create(conn, %Consumer{
             durable_name: consumer_name,
             deliver_subject: inbox,
             stream_name: stream,
             ack_policy: :none,
             max_ack_pending: -1,
             max_deliver: 1
           }) do
      {:ok, {sub, consumer_name}}
    end
  end
end
```

This way it cleans itself up. It also gives application developers the option to control the lifecycle and supervision of it if they want. You can still keep the convenience functions:

```elixir
def watch(conn, bucket_name, handler) do
  Jetstream.API.KV.Watcher.start_link(conn: conn, bucket_name: bucket_name, handler: handler)
end

def unwatch(pid) do
  Jetstream.API.KV.Watcher.stop(pid)
end
```